common/perf-test-core/src/main/java/com/azure/perf/test/core/EventPerfTest.java

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.perf.test.core;
+
+import reactor.core.publisher.Mono;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Represents the abstraction of a Performance test class.
+ *
+ * <p>
+ * The performance test class needs to extend this class.
+ * To add any test setup and logic the test class should override {@link EventPerfTest#globalSetupAsync()}
+ * and {@link EventPerfTest#globalCleanupAsync()} methods .
+ * </p>
+ * @param <TOptions> the options configured for the test.
+ */
+public abstract class EventPerfTest<TOptions extends PerfStressOptions> extends PerfTestBase<TOptions> {
+
+    private final AtomicInteger completedOps;
+
+    private volatile boolean errorRaised;
+
+    private long startTime;
+    private Throwable throwable;
+
+    /**
+     * Creates an instance of performance test.
+     * @param options the options configured for the test.
+     * @throws IllegalStateException if SSL context cannot be created.
+     */
+    public EventPerfTest(TOptions options) {
+        super(options);
+        if (options.getTestProxies() != null && options.getTestProxies().size() > 0) {
+            throw new IllegalStateException("Test Proxies are not supported for Event Perf Tests.");
+        }
+        completedOps = new AtomicInteger(0);
+    }
+
+    /**
+     * Indicates an event was raised, and records its count internally.
+     */
+    public void eventRaised() {
+        completedOps.getAndIncrement();
+        lastCompletionNanoTime = System.nanoTime() - startTime;
+    }
+    /**
+     * Indicates an error was raised, and stops the performance test flow.
+     */
+    public void errorRaised(Throwable throwable) {
+        errorRaised = true;
+        lastCompletionNanoTime = System.nanoTime() - startTime;
+        this.throwable = throwable;
+    }
+
+    @Override
+    public void runAll(long endNanoTime) {
+        startTime = System.nanoTime();
+        completedOps.set(0);
+        errorRaised = false;
+        lastCompletionNanoTime = 0;
+        while (System.nanoTime() < endNanoTime) {
+            if (errorRaised) {
+                throw new RuntimeException(throwable);
+            }
+        }
```

Comment:
This is a tight loop which will use 100% CPU of one core, significantly impacting performance.  You will need to use some sort of "interruptable timer" or synchronization primitives to block the thread until either:

1. `endNanoTime` is reached
2. `errorRaised()` was called

In .NET, this is implemented by creating a `CancellationTokenSource` that is cancelled when either of above conditions is met, and then calling `await Task.Delay(Timeout.InfiniteTimeSpan, cts.Token)` which waits (while using very little CPU) until the token is cancelled.

https://github.com/Azure/azure-sdk-for-net/blob/main/common/Perf/Azure.Test.Perf/EventPerfTest.cs#L73

In Java, I think you could do something similar with `wait()` and `notify()`, but I'm not certain on the details.  You probably need to wrap the call to `wait()` in a `try/catch` and do something with the `InterruptedException`.  Or there might be better Java APIs to do this.

```java
public void errorRaised(Throwable throwable) {
    errorRaised = true;
    lastCompletionNanoTime = System.nanoTime() - startTime;
    this.throwable = throwable;

    // NEW CODE
    notify();
}
```

```suggestion
        wait(endNanoTime - startTime);        
        if (errorRaised) {
            throw new RuntimeException(throwable);
        }
```

Reply:
wait and notify pattern works well, updated the code.